src/main/java/teammates/common/util/Logger.java

```diff
@@ -18,38 +28,98 @@ public static Logger getLogger() {
     }
 
     /**
      * Logs a message at WARNING level.
-     *
-     * @see java.util.logging.Logger#warning(String)
      */
-    public void warning(String msg) {
-        log.warning(msg);
+    public void warning(String message) {
+        standardLog.warning(formatLogMessage(message, "WARNING"));
     }
 
     /**
      * Logs a message at SEVERE level.
-     *
-     * @see java.util.logging.Logger#severe(String)
      */
-    public void severe(String msg) {
-        log.severe(msg);
+    public void severe(String message) {
+        errorLog.severe(formatLogMessage(message, "ERROR"));
+    }
+
+    private String formatLogMessage(String message, String severity) {
+        if (Config.isDevServer()) {
+            return formatLogMessageForHumanDisplay(message);
+        }
+        return formatLogMessageForCloudLogging(message, severity);
+    }
+
+    private String formatLogMessageForHumanDisplay(String message) {
+        StringBuilder prefix = new StringBuilder();
+
+        StackTraceElement source = getLoggerSource();
+        if (source != null) {
+            prefix.append(source.getClassName()).append(':')
+                    .append(source.getMethodName()).append(':')
+                    .append(source.getLineNumber()).append(':');
+        }
+        prefix.append(' ');
+
+        if (RequestTracer.getRequestId() == null) {
+            return prefix.toString() + message;
+        }
+        return prefix.toString() + "[" + RequestTracer.getRequestId() + "] " + message;
+    }
+
+    private String formatLogMessageForCloudLogging(String message, String severity) {
+        return JsonUtils.toCompactJson(getBaseCloudLoggingPayload(message, severity));
+    }
+
+    private Map<String, Object> getBaseCloudLoggingPayload(String message, String severity) {
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("message", message);
+        payload.put("severity", severity);
+
+        StackTraceElement source = getLoggerSource();
+        if (source != null) {
+            Map<String, Object> sourceLocation = new HashMap<>();
+            sourceLocation.put("file", source.getClassName());
+            sourceLocation.put("line", source.getLineNumber());
+            sourceLocation.put("function", source.getMethodName());
+
+            payload.put("logging.googleapis.com/sourceLocation", sourceLocation);
+        }
+
+        if (RequestTracer.getRequestId() != null) {
+            String[] traceAndSpan = RequestTracer.getRequestId().split("/", 2);
+            payload.put("logging.googleapis.com/trace", "projects/" + Config.APP_ID + "/traces/" + traceAndSpan[0]);
+            if (traceAndSpan.length == 2) {
+                payload.put("logging.googleapis.com/spanId", traceAndSpan[1].split(";")[0]);
+            }
+        }
+
+        return payload;
+    }
+
+    private StackTraceElement getLoggerSource() {
+        StackTraceElement[] stes = Thread.currentThread().getStackTrace();
```

Comment:
Getting a `stackTrace()` for every log we do and sweeping through it for the first element is a performance hit. `stackTrace()` walks through the stack and composes an array from it, even if we only need one element. 

Instead of that we can use [StackWalker](https://docs.oracle.com/javase/9/docs/api/java/lang/StackWalker.html). It is also thread safe and is more optimized - it does not create huge memory structures that go unused.

I do not have metrics to back this up but I am pretty sure this is an improvement here.

Reply:
StackWalker is a Java 9 feature.

Reply:
Ah yes - well something to keep in view after we upgrade the java version